### PR TITLE
Fix for AuthorizationConfigControllerTestIT#roleShouldBeDeleted Test case

### LINKED ISSUE #1744: Fix AuthorizationConfigControllerTestIT.roleShouldBeDeleted
# Task Description

We need to fix this test against the `issues/1649` branch.

Failed Jenkins build [here](https://jenkins.carlspring.org/blue/organizations/jenkins/strongbox%2Fbuilds%2Fstrongbox/detail/PR-1676/108/tests)

The problem seems related with JG issue that returrns single object value for `Cardinality.set` properties that contains only one element in collection. As a workaround we currently always add empty string into property collection to have more then one element in collection. But with current implementation its still posible to have single empty string element in collection. This should be fixed, and if original entity collection was empty then there is no need to add emty sting into property collection, so property collection should be empty as well.
Also we should improve this workaround with custom `EntityTraversalDsl` method (see below tasks).

# Tasks

The following tasks will need to be carried out:
* [ ] create `<E> EntityTraversalDsl#property(String key, Set<E> value)` method
* [ ] use `EntityTraversalDsl#property(String key, Set<E> value)` method within `VertexEntityTraversalAdapter` classes instead of common `property` method with `Cardinality.set`

# Task Relationships

This task:
* Is a sub-task of: #1649

# Help

* [Our chat](https://chat.carlspring.org/)
* Points of contact:
  * @carlspring
  * @sbespalov
  * @steve-todorov

### PR DESCRIPTION
**Pull request by anki2189**
# Pull Request Description

This pull request closes #1744 

# Acceptance Test

* [ ] Building the code with `mvn clean install -Dintegration.tests` still works.
* [ ] Running `mvn spring-boot:run` in the `strongbox-web-core` still starts up the application correctly.
* [ ] Building the code and running the `strongbox-distribution` from a `zip` or `tar.gz` still works.
* [ ] The tests in the [`strongbox-web-integration-tests`](https://github.com/strongbox/strongbox-web-integration-tests/) still run properly.

# Questions

* Does this pull request break backward compatibility? 
  * [ ] Yes
  * [x] No

* Does this pull request require other pull requests to be merged first? 
  * [ ] Yes, please see #...
  * [x] No

* Does this require an update of the documentation?
  * [ ] Yes, please see strongbox/strongbox-docs#{PR_NUMBER}
  * [x] No

### CODE REVIEW

**Comment by anki2189** [comment]
@sbespalov Pls review changes.

File: strongbox-data-service/src/main/java/org/carlspring/strongbox/gremlin/dsl/EntityTraversalDsl.java [hunk 1]
```diff
@@ -112,4 +115,24 @@ public String toString()
                                                                              ((Element) t.get()).property("uuid")
                                                                                                 .orElse("null"))));
     }
+    
+    default <E> GraphTraversal<S, E> property(final String key,
+                                              final Set<String> values,
+                                              GraphTraversal<S, E> graphTraversal)
```
Comment (sbespalov) [inline strongbox-data-service/src/main/java/org/carlspring/strongbox/gremlin/dsl/EntityTraversalDsl.java:121]: @anki2189 `graphTraversal` is superfluous here

File: strongbox-data-service/src/main/java/org/carlspring/strongbox/gremlin/dsl/EntityTraversalDsl.java [hunk 2]
```diff
@@ -112,4 +115,24 @@ public String toString()
...
+    {
+
+        if (!CollectionUtils.isEmpty(values))
```
Comment (sbespalov) [inline strongbox-data-service/src/main/java/org/carlspring/strongbox/gremlin/dsl/EntityTraversalDsl.java:124]:
```suggestion
        if (CollectionUtils.isEmpty(values))
        {
           return identity();
        }
```

File: strongbox-data-service/src/main/java/org/carlspring/strongbox/gremlin/dsl/EntityTraversalDsl.java [hunk 3]
```diff
@@ -112,4 +115,24 @@ public String toString()
...
+        {
+            if (values.size() == 1)
```
Comment (sbespalov) [inline strongbox-data-service/src/main/java/org/carlspring/strongbox/gremlin/dsl/EntityTraversalDsl.java:126]: @anki2189 why this condition needed?

File: strongbox-security/strongbox-user-management/src/test/java/org/carlspring/strongbox/repositories/UserRepositoryTest.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+import org.springframework.test.context.ContextConfiguration;
+
+import com.google.common.collect.Sets;
+
+/**
+ * @author ankit.tomar
+ */
+@SpringBootTest
+@ActiveProfiles(profiles = "test")
+@ContextConfiguration(classes = { DataServiceConfig.class,
```
Comment (sbespalov) [inline strongbox-security/strongbox-user-management/src/test/java/org/carlspring/strongbox/repositories/UserRepositoryTest.java:33]:
```suggestion
@ContextConfiguration(classes = RepositoriesTestConfig.class)
```

File: strongbox-security/strongbox-user-management/src/test/java/org/carlspring/strongbox/repositories/UserRepositoryTest.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+    @BeforeEach
+    public void before()
+    {
+        saveUserWithRoles("test-user1", Sets.newHashSet(SystemRole.ADMIN.name(), SystemRole.ARTIFACTS_MANAGER.name()));
+        saveUserWithRoles("test-user2", Sets.newHashSet(SystemRole.ADMIN.name(), "CUSTOM_ROLE"));
+    }
+
+    @Test
+    public void testFindAllUsers()
```
Comment (sbespalov) [inline strongbox-security/strongbox-user-management/src/test/java/org/carlspring/strongbox/repositories/UserRepositoryTest.java:49]: @anki2189 pls add a `@Transactional` annotation here as well.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/gremlin/adapters/GenericArtifactCoordinatesAdapter.java [hunk 6]
```diff
@@ -29,7 +29,7 @@
  * @author sbespalov
  */
 @Component
-public class GenericArtifactCoordinatesArapter extends VertexEntityTraversalAdapter<GenericArtifactCoordinates>
+public class GenericArtifactCoordinatesAdapter extends VertexEntityTraversalAdapter<GenericArtifactCoordinates>
```
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/gremlin/adapters/GenericArtifactCoordinatesAdapter.java:32]: :+1:

File: strongbox-data-service/src/main/java/org/carlspring/strongbox/gremlin/dsl/EntityTraversalDsl.java (hunk 1 above)
Comment (anki2189) [reply strongbox-data-service/src/main/java/org/carlspring/strongbox/gremlin/dsl/EntityTraversalDsl.java:121]: Then how can we get the same `GraphTraversal` instance on which other properties were set. As per my understanding it is working like chaining of properties

Comment (sbespalov) [reply strongbox-data-service/src/main/java/org/carlspring/strongbox/gremlin/dsl/EntityTraversalDsl.java:121]: > Then how can we get the same GraphTraversal instance on which other properties were set. 

this is `this` instance. you can have a look how other methods implemented here and there is also some [documentation](https://tinkerpop.apache.org/docs/current/reference/#dsl) on this.

File: strongbox-data-service/src/main/java/org/carlspring/strongbox/gremlin/dsl/EntityTraversalDsl.java (hunk 3 above)
Comment (anki2189) [reply strongbox-data-service/src/main/java/org/carlspring/strongbox/gremlin/dsl/EntityTraversalDsl.java:126]: This condition is added to handle single element Set use case, which will not work if we will not add empty string.

Comment (sbespalov) [reply strongbox-data-service/src/main/java/org/carlspring/strongbox/gremlin/dsl/EntityTraversalDsl.java:126]: I think we can omit this condition and always add empty element here.

File: strongbox-security/strongbox-user-management/src/test/java/org/carlspring/strongbox/repositories/UserRepositoryTest.java (hunk 5 above)
Comment (anki2189) [reply strongbox-security/strongbox-user-management/src/test/java/org/carlspring/strongbox/repositories/UserRepositoryTest.java:49]: Done.

File: strongbox-security/strongbox-user-management/src/test/java/org/carlspring/strongbox/repositories/UserRepositoryTest.java (hunk 4 above)
Comment (anki2189) [reply strongbox-security/strongbox-user-management/src/test/java/org/carlspring/strongbox/repositories/UserRepositoryTest.java]: Done.

File: strongbox-data-service/src/main/java/org/carlspring/strongbox/gremlin/dsl/EntityTraversalDsl.java (hunk 3 above)
Comment (anki2189) [reply strongbox-data-service/src/main/java/org/carlspring/strongbox/gremlin/dsl/EntityTraversalDsl.java]: Changed.

File: strongbox-data-service/src/main/java/org/carlspring/strongbox/gremlin/dsl/EntityTraversalDsl.java (hunk 1 above)
Comment (anki2189) [reply strongbox-data-service/src/main/java/org/carlspring/strongbox/gremlin/dsl/EntityTraversalDsl.java:121]: @sbespalov Can you pls change and push the changes , I am little bit confused.

File: strongbox-security/strongbox-user-management/src/test/java/org/carlspring/strongbox/repositories/UserRepositoryTest.java (hunk 4 above)
Comment (anki2189) [reply strongbox-security/strongbox-user-management/src/test/java/org/carlspring/strongbox/repositories/UserRepositoryTest.java:34]: This change is breaking other tests. Also `RepositoriesTestConfig` is itself importing `DataServiceConfig.class`.

**Comment by sbespalov** [comment]
@anki2189 merged, thanks

**Pull request merged**